### PR TITLE
fix: Speed up Trivia based Analyzers

### PR DESCRIPTION
**Pull request by ynsehoornenborg**
Experiments to see if the Trivia based Analyzers can be speed up.

### CODE REVIEW

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ProhibitDynamicKeywordAnalyzer.cs [hunk 1]
```diff
@@ -9,49 +10,94 @@
 namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
 {
 	[DiagnosticAnalyzer(LanguageNames.CSharp)]
-	public class ProhibitDynamicKeywordAnalyzer : SingleDiagnosticAnalyzer<IdentifierNameSyntax, ProhibitDynamicKeywordSyntaxNodeAction>
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ProhibitDynamicKeywordAnalyzer.cs:12]: It's hard to follow the diffs. What's happening in this class?

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAssignmentInConditionAnalyzer.cs [hunk 2]
```diff
@@ -37,8 +38,7 @@ public override void Initialize(AnalysisContext context)
 
 		private void AnalyzeIfStatement(SyntaxNodeAnalysisContext context)
 		{
-			GeneratedCodeDetector generatedCodeDetector = new();
-			if (generatedCodeDetector.IsGeneratedCode(context))
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAssignmentInConditionAnalyzer.cs:41]: Was moving the constructor a noticeable improvement? (I do think we can safely remove most of the GeneratedCodeDetector logic, as it's redundant with ConfigurationFlags.None.)

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/NoRegionsInMethodAnalyzer.cs [hunk 3]
```diff
@@ -28,12 +26,11 @@ public class NoRegionsInMethodSyntaxNodeAction : SyntaxNodeAction<MethodDeclarat
 		public override void Analyze()
 		{
 			// Specifying Span instead of FullSpan correctly excludes trivia before or after the method
-			var descendants = Node.DescendantNodes(Node.Span, null, descendIntoTrivia: true);
+			var descendants = Node.DescendantNodes(Node.Span, null, descendIntoTrivia: true).OfType<DirectiveTriviaSyntax>();
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/NoRegionsInMethodAnalyzer.cs:29]: We're calling this twice?

Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/NoRegionsInMethodAnalyzer.cs:29]: The performance gain here, is that the filtering is done earlier. That seemed to help speed up the foreach's that follow

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAssignmentInConditionAnalyzer.cs (hunk 2 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAssignmentInConditionAnalyzer.cs:41]: It seems like, but the measurement are unstable, see issue #402

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ProhibitDynamicKeywordAnalyzer.cs (hunk 1 above)
Comment (ynsehoornenborg) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/ProhibitDynamicKeywordAnalyzer.cs:12]: I transferred it from Trivia based into declaration based. This saved a ton of `SemanticModel` lookups, which are now only done when we know 90% sure its a `dynamic` type. Previsouly this analyzer dis a `SemanticModel` lookup for _every_ `Identifier` in the entire code.

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/NoRegionsInMethodAnalyzer.cs (hunk 3 above)
Comment (bcollamore) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/NoRegionsInMethodAnalyzer.cs:29]: Ahhh I missed DirectiveTriviaSyntax vs RegionDirectiveTriviaSyntax.

**Review by bcollamore — APPROVED** [review]
